sklearn/utils/_encode.py

```diff
@@ -48,9 +66,85 @@ def _unique(values, *, return_inverse=False, return_counts=False):
     )
 
 
+def _unique_pandas(values, *, return_counts=False):
+    if return_counts:
+        value_counts = values.value_counts(dropna=False, sort=False)
+        # sort categorical columns in lexical order to be consistent with order
+        # obtained when sorting after conversion to numpy array
+        try:
+            value_counts.index = value_counts.index.reorder_categories(
+                value_counts.index.categories.sort_values()
+            )
+        except AttributeError:
+            pass
+        value_counts = value_counts.sort_index()
+        return value_counts.index.to_numpy(), value_counts.to_numpy()
+    unique = values.unique()
+    # unique returns a NumpyExtensionArray for extension dtypes and a numpy
+    # array for other dtypes
+    if hasattr(unique, "sort_values"):
+        # sort categorical columns in lexical order to be consistent with order
+        # obtained when sorting after conversion to numpy array
+        try:
+            unique = unique.reorder_categories(unique.categories.sort_values())
```

Comment:
I do not see the need to `reorder_categories` here if we end up running `unique.sort_values()` afterwards. Can you provide an example where this is required?